### ISSUE TITLE
Intermittent NumberFormatException

### ISSUE THREAD

**Issue by damithc** [reposense/RepoSense#1601]
See #1600 #1599 

Perhaps we should add a retry mechanism as a temporary workaround, until we are able to pin down the exact cause?

**Comment by gerhean** [comment]
```
java.lang.NumberFormatException: empty String
at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842)
at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
at java.base/java.lang.Double.parseDouble(Double.java:543)
at java.base/java.text.DigitList.getDouble(DigitList.java:169)
at java.base/java.text.DecimalFormat.parse(DecimalFormat.java:2126)
at java.base/java.text.SimpleDateFormat.subParse(SimpleDateFormat.java:1933)
at java.base/java.text.SimpleDateFormat.parse(SimpleDateFormat.java:1541)
at java.base/java.text.DateFormat.parse(DateFormat.java:393)
at reposense.commits.CommitInfoAnalyzer.analyzeCommit(CommitInfoAnalyzer.java:92)
at reposense.commits.CommitInfoAnalyzer.lambda$analyzeCommits$0(CommitInfoAnalyzer.java:71)
```

This should actually be simple to fix.

Error is due to only catching ParseException but not NumberFormatException here.

https://github.com/reposense/RepoSense/blob/0c605e0e7fd77347c5cbd46b4783b1b8b7485699/src/main/java/reposense/commits/CommitInfoAnalyzer.java#L91-L95

Can be easily made to go away by just catching NumberFormatException.

**Comment by damithc** [comment]
> No idea why this error only surfaces now given that this file has not been changed recently, might signal that something else is wrong.

Yes, something else could be wrong. Note that this error is intermittent. It goes away upon re-running.

**Comment by gerhean** [comment]
Actually there's already something mentioned in the slack chat, just that nobody is addressing right now.
> I believe PR #1565 is the one that causes the bug

**Comment by dcshzj** [comment]
I did some investigations and it turns out that the problem is due to concurrency, as [explained in this StackOverflow question](https://stackoverflow.com/questions/63500822/returning-numberformatexception-form-simpledateformat-in-java-code). Simply put, `SimpleDateFormat` is not thread-safe, and this happens to us because we process the commits in parallel, which causes multiple threads to call the formatter at the same time. Hence, we see this error happen once in a while, due to the difficulty of debugging concurrency issues, and it only occurred recently after #1451 was merged.

I propose to switch to using the `java.time` package recommended by the answer in the above linked question. This would also save us a lot of headache when handling all time and date related issues throughout the codebase (especially #816 in the past).
